lib/ansible/modules/pip.py

```diff
@@ -611,6 +613,61 @@ def setup_virtualenv(module, env, chdir, out, err):
     return out, err, cmd
 
 
+def _normalize_vcs_packages(module: AnsibleModule,
+                            package_list: list[Package],
+                            pip: list[str]) -> list[Package]:
+    """Return list[Package] where all vcs URL packages are converted to actual package names"""
+    vcs_packages = [str(pkg) for pkg in package_list if _is_vcs_url(str(pkg))]
+
+    if not vcs_packages:
+        return package_list
+
+    other_packages = [pkg for pkg in package_list if not _is_vcs_url(str(pkg))]
+
+    # First, install as a dry-run report to get JSON output
+    frozen_env = os.environ.copy()
+    os.environ['TTY_COMPATIBLE'] = "0"
+    if os.environ.get('FORCE_COLOR'):
+        os.environ.pop('FORCE_COLOR')
+    rc, out, err = module.run_command([*pip, 'install', '--dry-run', '--ignore-installed', '--quiet', '--report=-', *vcs_packages])
+
+    if rc == 0:
+        # If it succeeds, handle the report
+        report = json.loads(out)
+        package_objects = [Package(install_report['metadata']['name'], version_string=install_report['metadata']['version'])
+                           for install_report in report['install']]
+    else:
+        # Else, if that fails due to --dry-run not being present in pip
+        # versions older than 22.2 (ex: Ubuntu 2204), fallback to using
+        # the potentially faulty pip download method and warn the user.
+
+        pip_downloads_dir = Path(module.tmpdir) / 'pipdownloads'
+        if pip_downloads_dir.exists():
+            shutil.rmtree(pip_downloads_dir)
+            pip_downloads_dir.mkdir()
+
+        rc, out, err = module.run_command([*pip, 'download', f'--dest={pip_downloads_dir}', '--no-deps', *vcs_packages])
+
+        if rc != 0:
+            # If it fails, just dump the error
+            os.environ.update(frozen_env)
+            module.fail_json(rc=rc, msg=out, err=err)
+
+        # Only from this point onward can we be sure that we're using the fallback, warn now.
+        module.warn("Using check_mode with vcs packages is potentially error prone on pip versions <22.2")
+
+        out_lines = out.splitlines()
+        saved_files = [Path(line.split(" ")[-1]).name for line in out_lines if 'Saved' in line]
+        package_names = [Path(name).stem for name in saved_files]
+
+        package_objects = [Package('-'.join(pkg.split('-')[:-1]),
+                                   version_string=pkg.split('-')[-1])
+                           for pkg in package_names]
+
+    os.environ.update(frozen_env)
+    return other_packages + package_objects
```

Comment:
Moving the computation from above and unwinding the generators is possible as follows:
```suggestion
    other_packages = (
        pkg for pkg in package_list
        if str(pkg) not in vcs_packages
    )
    return [*other_packages, *package_objects]
```

Reply:
+2 on generators, I always thought that the (x for x in y) was some kind of "tuple comprehension" -- whatever that would mean.

Reply:
@pkingstonxyz yeah, the official name of this syntax is generator expressions. When people say "generators" they usually mean "generator functions" (with `yield`). But most of the time, this difference is irrelevant.